File: iosApp/iosApp/Screens/SchoolPresenceRow.swift
import SwiftUI
import Shared

/// "Estoy aquí": quién está presente en esta escuela ahora mismo, con un
/// botón para marcarte tú también. Tocar el grupo de avatares/contador abre
/// la lista completa (hoja); desde ahí, tocar a alguien abre `ChatView`
/// (chat 1-a-1 ya existente). El icono de chat abre `SchoolChatView`, el
/// tablón abierto del sitio.
@MainActor
final class SchoolPresenceViewModel: ObservableObject {
    @Published var people: [SchoolPresence] = []
    @Published var iAmHere = false
    @Published var loading = false
    // Antes se tragaba cualquier fallo con `try?` sin decir nada — "Ya no
    // estoy" parecía no hacer nada y no había forma de saber por qué
    // (Álvaro, 2026-09-03). Ahora, si algo falla de verdad, se ve en pantalla.
    @Published var errorText: String?

    private let schoolId: String
    private let myUid: String?
    private let getPresence = AppDependencies.shared.container.getSchoolPresence
    private let markPresence = AppDependencies.shared.container.markSchoolPresence
    private let clearPresence = AppDependencies.shared.container.clearSchoolPresence

    init(schoolId: String, myUid: String?) {
        self.schoolId = schoolId
        self.myUid = myUid
    }

    func load() async {
        // Público: sin sesión también se ve quién hay, igual que el resto de la ficha.
        do {
            let list = try await getPresence.execute(schoolId: schoolId)
            people = list
            iAmHere = myUid.map { uid in list.contains { $0.uid == uid } } ?? false
        } catch {
            errorText = "No se pudo cargar quién hay aquí: \(error.localizedDescription)"
        }
    }

    func toggle() {
        guard !loading else { return }
        loading = true
        errorText = nil
        Task {
            defer { loading = false }
            do {
                if iAmHere {
                    try await clearPresence.execute(schoolId: schoolId)
                } else {
                    _ = try await markPresence.execute(schoolId: schoolId)
                }
            } catch {
                errorText = "\(iAmHere ? "No se pudo quitar" : "No se pudo marcar") la presencia: \(error.localizedDescription)"
            }
            await load()
        }
    }
}

/// Identifica con quién abrir el chat 1-a-1. `Identifiable` para poder usar
/// `navigationDestination(item:)` — el `NavigationLink(isActive:)` de antes
/// (en `.background()`) se comía el toque del botón "Ya no estoy" al
/// compartir zona de gesto con el resto de la fila (Álvaro, 2026-09-03:
/// el botón no respondía en TestFlight). `navigationDestination(item:)` no
/// necesita una vista fantasma en el árbol, así que no interfiere.
private struct ChatTarget: Identifiable, Hashable {
    let uid: String
    let name: String
    var id: String { uid }
}

struct SchoolPresenceRow: View {
    let schoolId: String
    let schoolName: String
    let myUid: String?
    @StateObject private var vm: SchoolPresenceViewModel
    @State private var showPrivacyNote = false
    @State private var openChatFor: ChatTarget?
    @State private var showSchoolChat = false
    @State private var showAllPresent = false

    init(schoolId: String, schoolName: String, myUid: String?) {
        self.schoolId = schoolId
        self.schoolName = schoolName
        self.myUid = myUid
        _vm = StateObject(wrappedValue: SchoolPresenceViewModel(schoolId: schoolId, myUid: myUid))
    }

    var body: some View {
        VStack(spacing: 0) {
            Group {
                if !vm.people.isEmpty || vm.iAmHere {
                    content
                } else {
                    // Nadie presente: la misma barra fina, con chat + botón a la derecha.
                    HStack {
                        Spacer()
                        chatButton
                        markButton
                    }
                    .padding(.horizontal, 16)
                    .padding(.vertical, 5)
                    .frame(height: 34)
                    .background(Cumbre.paper2)
                    .overlay(Rectangle().frame(height: 1).foregroundStyle(Cumbre.rule), alignment: .bottom)
                }
            }
            if let err = vm.errorText {
                Text(err)
                    .font(.system(size: 10.5))
                    .foregroundStyle(Cumbre.bad)
                    .padding(.horizontal, 16).padding(.vertical, 6)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .background(Cumbre.bad.opacity(0.1))
            }
        }
        .task { await vm.load() }
        .sheet(isPresented: $showPrivacyNote) {
            privacySheet
        }
        .navigationDestination(item: $openChatFor) { target in
            ChatView(otherUid: target.uid, otherName: target.name)
        }
        .navigationDestination(isPresented: $showSchoolChat) {
            SchoolChatView(schoolId: schoolId, schoolName: schoolName)
        }
        .sheet(isPresented: $showAllPresent) {
            allPresentSheet
        }
    }

    private var content: some View {
        HStack(spacing: 6) {
            // Un toque en cualquier parte del grupo de avatares abre la lista
            // completa — con más de 4 presentes, tocar un avatar concreto en
            // la pila superpuesta es ambiguo (¿cuál de los que se tapan he
            // tocado?); la lista deja elegir con nombre y sin dudas.
            Button { showAllPresent = true } label: {
                HStack(spacing: -6) {
                    ForEach(vm.people.prefix(4), id: \.uid) { person in
                        AvatarCircle(url: person.photoUrl, size: 18)
                            .overlay(Circle().stroke(Cumbre.paper2, lineWidth: 1.5))
                    }
                }
            }
            Button { showAllPresent = true } label: {
                Text(peopleLabel)
                    .font(.system(size: 11.5, weight: .semibold, design: .serif))
                    .foregroundStyle(Cumbre.ink)
            }
            Spacer()
            chatButton
            markButton
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 5)
        .frame(height: 34)
        .background(Cumbre.paper2)
        .overlay(Rectangle().frame(height: 1).foregroundStyle(Cumbre.rule), alignment: .bottom)
    }

    private var allPresentSheet: some View {
        NavigationStack {
            List(vm.people, id: \.uid) { person in
                Button {
                    guard person.uid != myUid else { return }
                    showAllPresent = false
                    openChatFor = ChatTarget(uid: person.uid, name: person.displayName ?? person.username ?? "Usuario")
                } label: {
                    HStack(spacing: 12) {
                        AvatarCircle(url: person.photoUrl, size: 36)
                        Text(person.uid == myUid ? "Tú" : (person.displayName ?? person.username ?? "Usuario"))
                            .font(.system(size: 15, design: .serif))
                            .foregroundStyle(Cumbre.ink)
                        Spacer()
                        if person.uid != myUid {
                            Image(systemName: "bubble.left").font(.system(size: 13)).foregroundStyle(Cumbre.ink3)
                        }
                    }
                }
                .disabled(person.uid == myUid)
            }
            .navigationTitle("\(vm.people.count) aquí ahora")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .navigationBarTrailing) {
                    Button("Cerrar") { showAllPresent = false }
                }
            }
        }
        .presentationDetents([.medium, .large])
    }

    private var chatButton: some View {
        Button { showSchoolChat = true } label: {
            Image(systemName: "bubble.left.and.bubble.right")
                .font(.system(size: 13))
                .foregroundStyle(Cumbre.ink2)
                .frame(width: 26, height: 26)
                .background(Cumbre.paper, in: Circle())
                .overlay(Circle().stroke(Cumbre.rule, lineWidth: 1))
        }
    }

    private var peopleLabel: String {
        vm.people.isEmpty ? "" : "\(vm.people.count) aquí ahora"
    }

    private var markButton: some View {
        Button {
            if vm.iAmHere {
                vm.toggle()
            } else {
                showPrivacyNote = true
            }
        } label: {
            HStack(spacing: 4) {
                Image(systemName: "mappin.circle.fill").font(.system(size: 11))
                Text(vm.iAmHere ? "Ya no estoy" : "Estoy aquí")
                    .font(.system(size: 11, weight: .bold, design: .serif))
            }
            .foregroundStyle(.white)
            .padding(.horizontal, 10).padding(.vertical, 5)
            .background(vm.iAmHere ? Cumbre.ink3 : Cumbre.terraFill)
            .clipShape(RoundedRectangle(cornerRadius: 7))
        }
        .disabled(vm.loading)
    }

    private var privacySheet: some View {
        VStack(alignment: .leading, spacing: 16) {
            Text("Al marcar \"Estoy aquí\"")
                .font(.system(size: 20, weight: .bold, design: .serif))
            Text("Cualquiera que abra esta escuela verá que estás aquí y podrá escribirte por chat — aunque tu perfil sea privado. Nadie podrá ver tu perfil completo si no te sigue. Se desactiva sola pasadas 10 horas, o puedes quitarla tú antes.")
                .font(.system(size: 15))
                .foregroundStyle(Cumbre.ink2)
            Button {
                showPrivacyNote = false
                vm.toggle()
            } label: {
                Text("Entendido, estoy aquí")
                    .font(.system(size: 15, weight: .bold, design: .serif))
                    .foregroundStyle(.white)
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, 12)
                    .background(Cumbre.terraFill)
                    .clipShape(RoundedRectangle(cornerRadius: 10))
            }
        }
        .padding(24)
        .presentationDetents([.medium])
    }
}

File: app/src/main/java/com/meteomontana/android/ui/components/SchoolPresenceRow.kt
package com.meteomontana.android.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Chat
import androidx.compose.material.icons.outlined.Place
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import coil.compose.AsyncImage
import com.meteomontana.android.domain.model.SchoolPresence
import com.meteomontana.android.domain.usecase.presence.ClearSchoolPresenceUseCase
import com.meteomontana.android.domain.usecase.presence.GetSchoolPresenceUseCase
import com.meteomontana.android.domain.usecase.presence.MarkSchoolPresenceUseCase
import com.meteomontana.android.ui.theme.Serif
import com.meteomontana.android.ui.theme.Spacing
import com.meteomontana.android.ui.theme.Terra
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class SchoolPresenceViewModel @Inject constructor(
    private val getPresence: GetSchoolPresenceUseCase,
    private val markPresence: MarkSchoolPresenceUseCase,
    private val clearPresence: ClearSchoolPresenceUseCase
) : ViewModel() {
    var people by mutableStateOf<List<SchoolPresence>>(emptyList())
        private set
    var iAmHere by mutableStateOf(false)
        private set
    var loading by mutableStateOf(false)
        private set
    // Antes se tragaba cualquier fallo en silencio — "Ya no estoy" parecía no
    // hacer nada y no había forma de saber por qué. Si algo falla, se ve.
    var errorText by mutableStateOf<String?>(null)
        private set

    fun load(schoolId: String, myUid: String?) {
        viewModelScope.launch {
            try {
                val list = getPresence.execute(schoolId)
                people = list
                iAmHere = myUid != null && list.any { it.uid == myUid }
            } catch (e: Exception) {
                errorText = "No se pudo cargar quién hay aquí: ${e.message}"
            }
        }
    }

    fun toggle(schoolId: String, myUid: String?) {
        if (loading) return
        loading = true
        errorText = null
        viewModelScope.launch {
            try {
                if (iAmHere) clearPresence.execute(schoolId) else markPresence.execute(schoolId)
            } catch (e: Exception) {
                errorText = "${if (iAmHere) "No se pudo quitar" else "No se pudo marcar"} la presencia: ${e.message}"
            }
            loading = false
            load(schoolId, myUid)
        }
    }
}

/**
 * "Estoy aquí": quién está presente en esta escuela ahora mismo, con un botón
 * para marcarte tú también. Fija bajo la cabecera, fuera del scroll — espejo
 * exacto de `SchoolPresenceRow.swift`.
 */
@Composable
fun SchoolPresenceRow(
    schoolId: String,
    schoolName: String,
    myUid: String?,
    onOpenChat: (uid: String, name: String) -> Unit,
    onOpenSchoolChat: () -> Unit,
    // Sube al deslizar hacia abajo en la ficha (pull-to-refresh) para pedir
    // otra vez quién hay — sin caché, es información en vivo (Álvaro,
    // 2026-09-04: mejor a demanda que sondear sola cada X segundos).
    refreshKey: Int = 0,
    viewModel: SchoolPresenceViewModel = hiltViewModel()
) {
    var showPrivacyNote by remember { mutableStateOf(false) }
    var showAllPresent by remember { mutableStateOf(false) }

    androidx.compose.runtime.LaunchedEffect(schoolId, myUid, refreshKey) {
        viewModel.load(schoolId, myUid)
    }

    Column {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .background(MaterialTheme.colorScheme.background)
                .padding(horizontal = Spacing.md, vertical = Spacing.xs)
                .height(34.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            if (viewModel.people.isNotEmpty()) {
                Row(
                    modifier = Modifier.clickable { showAllPresent = true },
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Box {
                        viewModel.people.take(4).forEachIndexed { idx, person ->
                            Box(
                                modifier = Modifier
                                    .padding(start = (idx * 12).dp)
                                    .size(18.dp)
                                    .clip(CircleShape)
                                    .border(1.5.dp, MaterialTheme.colorScheme.background, CircleShape)
                            ) {
                                PresenceAvatar(url = person.photoUrl, size = 18.dp)
                            }
                        }
                    }
                    Spacer(Modifier.padding(start = (viewModel.people.take(4).size * 12 + 6).dp))
                    Text(
                        "${viewModel.people.size} aquí ahora",
                        style = MaterialTheme.typography.labelMedium.copy(fontFamily = Serif, fontWeight = FontWeight.SemiBold),
                        color = MaterialTheme.colorScheme.onBackground
                    )
                }
            }
            Spacer(Modifier.weight(1f))
            PresenceChatButton(onClick = onOpenSchoolChat)
            Spacer(Modifier.padding(start = Spacing.xs))
            PresenceMarkButton(
                iAmHere = viewModel.iAmHere,
                loading = viewModel.loading,
                onClick = {
                    if (viewModel.iAmHere) viewModel.toggle(schoolId, myUid)
                    else showPrivacyNote = true
                }
            )
        }
        androidx.compose.material3.HorizontalDivider(color = MaterialTheme.colorScheme.outline, thickness = 1.dp)
        viewModel.errorText?.let { err ->
            Text(
                err,
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.error,
                modifier = Modifier
                    .fillMaxWidth()
                    .background(MaterialTheme.colorScheme.error.copy(alpha = 0.1f))
                    .padding(horizontal = Spacing.md, vertical = Spacing.xs)
            )
        }
    }

    if (showPrivacyNote) {
        PresencePrivacySheet(
            onConfirm = {
                showPrivacyNote = false
                viewModel.toggle(schoolId, myUid)
            },
            onDismiss = { showPrivacyNote = false }
        )
    }

    if (showAllPresent) {
        PresenceAllSheet(
            people = viewModel.people,
            myUid = myUid,
            onDismiss = { showAllPresent = false },
            onOpenChat = { uid, name ->
                showAllPresent = false
                onOpenChat(uid, name)
            }
        )
    }
}

@Composable
private fun PresenceAvatar(url: String?, size: androidx.compose.ui.unit.Dp) {
    if (url != null) {
        AsyncImage(model = url, contentDescription = null,
            modifier = Modifier.size(size).clip(CircleShape))
    } else {
        Box(
            modifier = Modifier.size(size).clip(CircleShape)
                .background(MaterialTheme.colorScheme.surfaceVariant)
        )
    }
}

@Composable
private fun PresenceChatButton(onClick: () -> Unit) {
    Box(
        modifier = Modifier
            .size(26.dp)
            .clip(CircleShape)
            .background(MaterialTheme.colorScheme.surface)
            .border(1.dp, MaterialTheme.colorScheme.outline, CircleShape)
            .clickable(onClick = onClick),
        contentAlignment = Alignment.Center
    ) {
        Icon(Icons.Outlined.Chat, contentDescription = "Chat de la escuela",
            tint = MaterialTheme.colorScheme.onSurfaceVariant, modifier = Modifier.size(13.dp))
    }
}

@Composable
private fun PresenceMarkButton(iAmHere: Boolean, loading: Boolean, onClick: () -> Unit) {
    Row(
        modifier = Modifier
            .clip(RoundedCornerShape(7.dp))
            .background(if (iAmHere) MaterialTheme.colorScheme.onSurfaceVariant else Terra)
            .clickable(enabled = !loading, onClick = onClick)
            .padding(horizontal = 10.dp, vertical = 5.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Icon(Icons.Outlined.Place, contentDescription = null,
            tint = Color.White, modifier = Modifier.size(11.dp))
        Spacer(Modifier.padding(start = 4.dp))
        Text(
            if (iAmHere) "Ya no estoy" else "Estoy aquí",
            style = MaterialTheme.typography.labelSmall.copy(fontFamily = Serif, fontWeight = FontWeight.Bold),
            color = Color.White
        )
    }
}

@Composable
private fun PresencePrivacySheet(onConfirm: () -> Unit, onDismiss: () -> Unit) {
    Dialog(onDismissRequest = onDismiss, properties = DialogProperties(usePlatformDefaultWidth = false)) {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .padding(Spacing.lg)
                .clip(RoundedCornerShape(16.dp))
                .background(MaterialTheme.colorScheme.background)
                .padding(Spacing.lg)
        ) {
            Column {
                Text(
                    "Al marcar \"Estoy aquí\"",
                    style = MaterialTheme.typography.titleLarge.copy(fontFamily = Serif, fontWeight = FontWeight.Bold),
                    color = MaterialTheme.colorScheme.onBackground
                )
                Spacer(Modifier.padding(top = Spacing.sm))
                Text(
                    "Cualquiera que abra esta escuela verá que estás aquí y podrá escribirte por chat — aunque tu perfil sea privado. Nadie podrá ver tu perfil completo si no te sigue. Se desactiva sola pasadas 10 horas, o puedes quitarla tú antes.",
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                Spacer(Modifier.padding(top = Spacing.md))
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(10.dp))
                        .background(Terra)
                        .clickable(onClick = onConfirm)
                        .padding(vertical = Spacing.md),
                    contentAlignment = Alignment.Center
                ) {
                    Text(
                        "Entendido, estoy aquí",
                        style = MaterialTheme.typography.bodyMedium.copy(fontFamily = Serif, fontWeight = FontWeight.Bold),
                        color = Color.White
                    )
                }
            }
        }
    }
}

@Composable
private fun PresenceAllSheet(
    people: List<SchoolPresence>,
    myUid: String?,
    onDismiss: () -> Unit,
    onOpenChat: (uid: String, name: String) -> Unit
) {
    Dialog(onDismissRequest = onDismiss, properties = DialogProperties(usePlatformDefaultWidth = false)) {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(topStart = 16.dp, topEnd = 16.dp))
                .background(MaterialTheme.colorScheme.background)
        ) {
            Column {
                Row(
                    modifier = Modifier.fillMaxWidth().padding(Spacing.md),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        "${people.size} aquí ahora",
                        style = MaterialTheme.typography.titleMedium.copy(fontFamily = Serif, fontWeight = FontWeight.Bold),
                        color = MaterialTheme.colorScheme.onBackground,
                        modifier = Modifier.weight(1f)
                    )
                    Text(
                        "Cerrar",
                        style = MaterialTheme.typography.bodyMedium,
                        color = Terra,
                        modifier = Modifier.clickable(onClick = onDismiss)
                    )
                }
                androidx.compose.material3.HorizontalDivider(color = MaterialTheme.colorScheme.outline, thickness = 1.dp)
                LazyColumn(modifier = Modifier.height(400.dp)) {
                    items(people) { person ->
                        val isMe = person.uid == myUid
                        Row(
                            modifier = Modifier
                                .fillMaxWidth()
                                .clickable(enabled = !isMe) {
                                    onOpenChat(person.uid, person.displayName ?: person.username ?: "Usuario")
                                }
                                .padding(horizontal = Spacing.md, vertical = Spacing.sm),
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            PresenceAvatar(url = person.photoUrl, size = 36.dp)
                            Spacer(Modifier.padding(start = Spacing.sm))
                            Text(
                                if (isMe) "Tú" else (person.displayName ?: person.username ?: "Usuario"),
                                style = MaterialTheme.typography.bodyLarge.copy(fontFamily = Serif),
                                color = MaterialTheme.colorScheme.onBackground,
                                modifier = Modifier.weight(1f)
                            )
                            if (!isMe) {
                                Icon(Icons.Outlined.Chat, contentDescription = null,
                                    tint = MaterialTheme.colorScheme.onSurfaceVariant, modifier = Modifier.size(16.dp))
                            }
                        }
                    }
                }
            }
        }
    }
}
